### PR TITLE
Add Caching to Tree Hashing Algorithm

### PR DESCRIPTION
**Pull request by shayzluf**
[Resolves] #1638 

---

# Making the tree hashing algorithm more efficient by caching branches  

Cashing struct and slice hashes help get faster execution of the hash function in ssz

added hash cache and made sszutil getHasher use the caching mechanism

### CODE REVIEW

File: proto/beacon/p2p/v1/messages.pb.go [hunk 1]
```diff
@@ -5,10 +5,9 @@ package ethereum_beacon_p2p_v1
 
 import (
 	fmt "fmt"
+	proto "github.com/gogo/protobuf/proto"
```
Comment (nisdas) [inline proto/beacon/p2p/v1/messages.pb.go]: Are these changes from goimports ? i think we have them fixed

File: shared/ssz/hash_cache.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	// ErrNotMarakleRoot will be returned when a cache object is not a marakle root
+	ErrNotMarakleRoot = errors.New("object is not a markle root")
```
Comment (nisdas) [inline shared/ssz/hash_cache.go]: its `MerkleRoot` not `MarakleRoot`

File: shared/ssz/hash_cache.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+		Name: "powchain_hash_cache_hit",
+		Help: "The number of hash requests that are present in the cache.",
+	})
+	hashCacheSize = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "powchain_hash_cache_size",
+		Help: "The number of hashes in the block cache",
+	})
+)
+
+// markleRoot specifies the hash of data in a struct
```
Comment (nisdas) [inline shared/ssz/hash_cache.go]: Again its `MerkleRoot`

File: shared/ssz/hash_cache.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+	}
+
+	return mRoot.Hash.Hex(), nil
+}
+
+// newHashCache creates a new hash cache for storing/accessing root hashes from
+// memory.
+func newHashCache() *hashCacheS {
+	return &hashCacheS{
+		hashCache: cache.NewFIFO(hashKeyFn),
```
Comment (nisdas) [inline shared/ssz/hash_cache.go]: why is a fifo cache being used ? there is no ordering to whatever objects we are hashing.

File: shared/ssz/hash_cache.go [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+		return false, nil, ErrNotMarakleRoot
+	}
+
+	return true, hInfo, nil
+}
+
+// AddRetrieveTrieRoot adds a trie root to the cache. This method also trims the
+// least recently added root info if the cache size has reached the max cache
+// size limit.
+func (b *hashCacheS) AddRetrieveTrieRoot(val interface{}) ([32]byte, error) {
```
Comment (nisdas) [inline shared/ssz/hash_cache.go]: The naming is odd, `AddRetrieveTrieRoot` signifies two opposite actions by this method.

File: shared/ssz/hash_cache.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+		}
+		// Right-pad with 0 to make 32 bytes long, if necessary
+		paddedOutput = bytesutil.ToBytes32(output)
+		b.AddRoot(bytesutil.ToBytes32(hs), paddedOutput[:])
+	}
+
+	return paddedOutput, nil
+}
+
+// AddRetriveMarkleRoot adds a mrakle object to the cache. This method also trims the
```
Comment (nisdas) [inline shared/ssz/hash_cache.go]: again spelling mistake

File: shared/ssz/hash_cache.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+// least recently added root info if the cache size has reached the max cache
+// size limit.
+func (b *hashCacheS) AddRetriveMarkleRoot(byteSlice [][]byte) ([]byte, error) {
```
Comment (nisdas) [inline shared/ssz/hash_cache.go]: The naming is again confusing, and it indicates its doing the same exact thing as the above function

File: shared/ssz/hash_cache.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+// hashCacheS struct with one queue for looking up by hash.
+type hashCacheS struct {
+	hashCache *cache.FIFO
+}
+
+// RootByHash fetches Root by its hash. Returns true with a
+// reference to the root if exists. Otherwise returns false, nil.
+func (b *hashCacheS) RootByHash(h common.Hash) (bool, *root, error) {
```
Comment (nisdas) [inline shared/ssz/hash_cache.go]: So how would this work ? if you retrieve the root by its hash, you would have to generate the root in the first place which defeats the purpose of caching it

File: shared/ssz/hash_cache_test.go [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+		})
+
+	}
+
+}
+
+// EncodeDepositData converts a deposit input proto into an a byte slice
+// of Simple Serialized deposit input followed by 8 bytes for a deposit value
+// and 8 bytes for a unix timestamp, all in LittleEndian format.
+func EncodeDepositData(
```
Comment (nisdas) [inline shared/ssz/hash_cache_test.go]: Why is this here ? There should be no reason we have this in the ssz library.

File: shared/ssz/hash_cache_test.go [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+	depositData = append(depositData, value...)
+	depositData = append(depositData, timestamp...)
+	depositData = append(depositData, encodedInput...)
+
+	return depositData, nil
+}
+
+// generateInitialSimulatedDeposits generates initial deposits for creating a beacon state in the simulated
+// backend based on the yaml configuration.
+func generateInitialSimulatedDeposits(numDeposits uint64) ([]*pb.Deposit, []*bls.SecretKey, error) {
```
Comment (nisdas) [inline shared/ssz/hash_cache_test.go]: Again this makes no sense, why are we adding a function from the simulated backend here.

File: shared/ssz/hash_cache_test.go [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+		if err != nil {
+			return nil, nil, fmt.Errorf("could not encode genesis block deposits: %v", err)
+		}
+		deposits[i] = &pb.Deposit{DepositData: depositData}
+		privKeys[i] = priv
+	}
+	return deposits, privKeys, nil
+}
+
+func TestDepositsBLSHash(t *testing.T) {
```
Comment (nisdas) [inline shared/ssz/hash_cache_test.go]: What is the point of this ? What does this have to do with ssz.

File: shared/ssz/hash_cache.go (hunk 4 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: you are right

File: shared/ssz/hash_cache.go (hunk 5 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: it retrieves from cache if this object is there and calculate the root and add it to cache when its missing

File: shared/ssz/hash_cache.go (hunk 8 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: Ok this is an important question. it retrieves the root by the hash of encoded form of the object. While the regular flow is to hash each branch of the object by itself and then hash them all into one

File: shared/ssz/hash_cache_test.go (hunk 10 above)
Comment (shayzluf) [reply shared/ssz/hash_cache_test.go]: couldn't import simulated backend because of circular dependency so that is what I chose to do in order to generate data that is similar for what we are going to process

File: proto/beacon/p2p/v1/messages.pb.go (hunk 1 above)
Comment (shayzluf) [reply proto/beacon/p2p/v1/messages.pb.go]: I ran gazzlle --fix and update-go-pbs.sh
I didn't add this line

File: shared/ssz/hash_cache.go (hunk 5 above)
Comment (nisdas) [reply shared/ssz/hash_cache.go]: Sure but add and retrieve mean two very different things, when I read the method I do not have a good idea what it does because the naming is confusing.

File: proto/beacon/p2p/v1/messages.pb.go (hunk 1 above)
Comment (nisdas) [reply proto/beacon/p2p/v1/messages.pb.go]: Thats fine then

File: shared/ssz/hash_cache.go (hunk 8 above)
Comment (nisdas) [reply shared/ssz/hash_cache.go]: In that case could you update the method decription to describe more accurately what the method does. When you say `Root by its hash` it reads as we hash the root of the encoded object, not that we hash the encoded object.

File: shared/ssz/hash_cache_test.go (hunk 10 above)
Comment (nisdas) [reply shared/ssz/hash_cache_test.go]: imo this isnt a good place for this, the ssz package is a shared package which might be used by other projects if they want to use ssz. Bringing eth 2.0 specific logic here isnt good, we want to keep it as self contained as possible. If you want to test how we tree hash proto files, you can just create a proto object with junk data. There is no reason to use simulated backend here, for what is a serialization library.

File: shared/ssz/hash_cache.go (hunk 5 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: got you. I hope I got the naming better this time

File: shared/ssz/hash_cache.go (hunk 7 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: fixed

File: shared/ssz/hash_cache.go (hunk 8 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: fixed

File: shared/ssz/hash_cache_test.go (hunk 10 above)
Comment (shayzluf) [reply shared/ssz/hash_cache_test.go]: removed

File: shared/ssz/hash_cache.go (hunk 4 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: replaced

File: shared/params/config.go [hunk 12]
```diff
@@ -22,6 +22,7 @@ type BeaconChainConfig struct {
 	ValidatorPrivkeyFileName     string // ValidatorPrivKeyFileName specifies the string name of a validator private key file.
 	WithdrawalPrivkeyFileName    string // WithdrawalPrivKeyFileName specifies the string name of a withdrawal private key file.
 	BLSPubkeyLength              int    // BLSPubkeyLength defines the expected length of BLS public keys in bytes.
+	HashCacheSize                int    // HashCacheSize defines the size of object hashes that are cached
```
Comment (terenc3t) [inline shared/params/config.go]:
```suggestion
	HashCacheSize                int    // HashCacheSize defines the size of object hashes that's cached.
```

File: shared/params/config.go [hunk 13]
```diff
@@ -112,6 +113,7 @@ var defaultBeaconConfig = &BeaconChainConfig{
 	ValidatorPrivkeyFileName:     "/validatorprivatekey",
 	WithdrawalPrivkeyFileName:    "/shardwithdrawalkey",
 	BLSPubkeyLength:              96,
+	HashCacheSize:                100000,
```
Comment (terenc3t) [inline shared/params/config.go]: why 100000?

File: shared/ssz/hash_cache.go [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+var (
+	// ErrNotMerkleRoot will be returned when a cache object is not a merkle root
```
Comment (terenc3t) [inline shared/ssz/hash_cache.go]:
```suggestion
	// ErrNotMerkleRoot will be returned when a cache object is not a merkle root.
```

File: shared/params/config.go (hunk 13 above)
Comment (shayzluf) [reply shared/params/config.go]: thought it wont be too aggressive on the memory. What would you suggest?

File: shared/ssz/hash_cache.go [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+	startTime := time.Now().UnixNano()
+	hs, err := hashedEncoding(rval)
+	if err != nil {
+		return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+	}
+	exists, fetchedInfo, err := b.RootByEncodedHash(bytesutil.ToBytes32(hs))
+	if err != nil {
+		return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+	}
+	fmt.Printf("encoding time: %v \n", time.Now().UnixNano()-startTime)
```
Comment (rauljordan) [inline shared/ssz/hash_cache.go]: Hi @shayzluf let's remove this

File: shared/ssz/hash_cache.go [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+		if err != nil {
+			return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+		}
+		startTime = time.Now().UnixNano()
+		output, err := sszUtils.hasher(rval)
+		fmt.Printf("hashing time: %v \n", time.Now().UnixNano()-startTime)
+		if err != nil {
+			return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+		}
+		// Right-pad with 0 to make 32 bytes long, if necessary
```
Comment (rauljordan) [inline shared/ssz/hash_cache.go]: Use punctuation in comments.

File: shared/ssz/hash_cache_test.go [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+
+		runMerkleHashTests(t, func(val [][]byte) ([]byte, error) {
+			return cache.MerkleHashCached(val)
+		})
+
+	}
+
+}
+
+type JunkObject struct {
```
Comment (rauljordan) [inline shared/ssz/hash_cache_test.go]: Please put this struct and helper functions at the top of the test file

File: shared/ssz/hash_cache_test.go [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+	return object
+}
+
+func TestBenchmarHashWithCache(t *testing.T) {
+
+	useCache = false
+
+	First := GenerateJunkObject(10000)
+
+	type tree struct {
```
Comment (rauljordan) [inline shared/ssz/hash_cache_test.go]: Do not define structs inside test functions

File: shared/ssz/hash_cache_test.go [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+		First  []*JunkObject
+		Second []*JunkObject
+	}
+	startTime := time.Now().UnixNano()
+	TreeHash(&tree{First: First, Second: First})
+	fmt.Printf("time it took without cache: %v \n", time.Now().UnixNano()-startTime)
```
Comment (rauljordan) [inline shared/ssz/hash_cache_test.go]: Do not use fmt to do benches, instead use `*testing.B` and go benchmarking tools. See here: https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go

File: shared/ssz/hash_cache_test.go (hunk 17 above)
Comment (shayzluf) [reply shared/ssz/hash_cache_test.go]: done

File: shared/ssz/hash_cache_test.go (hunk 19 above)
Comment (shayzluf) [reply shared/ssz/hash_cache_test.go]: Thanks

**Comment by prestonvanloon** [comment]
Any update on this? Still WIP? 

We can merge this if you add a feature flag to conditionally enable the new feature.

**Comment by shayzluf** [comment]
> Any update on this? Still WIP?
> 
> We can merge this if you add a feature flag to conditionally enable the new feature.
@prestonvanloon 
I added the feature flag but ssz is not being used in either beacon chain nor validator so the feature flag is not being used yet

File: shared/featureconfig/flags.go [hunk 20]
```diff
@@ -5,6 +5,11 @@ import (
 )
 
 var (
+	// CacheTreeHashFlag determines whether to cache tree hashes for ssz.
+	CacheTreeHashFlag = cli.BoolFlag{
+		Name:  "cache-tree-hash",
```
Comment (prestonvanloon) [inline shared/featureconfig/flags.go]: Please use the format `--enable-cache-tree-hash`

File: shared/ssz/hash.go [hunk 21]
```diff
@@ -51,6 +55,9 @@ func newHashError(msg string, typ reflect.Type) *hashError {
 }
 
 func makeHasher(typ reflect.Type) (hasher, error) {
+	if featureconfig.FeatureConfig().CacheTreeHash {
```
Comment (prestonvanloon) [inline shared/ssz/hash.go]:
```suggestion
	useCache = featureconfig.FeatureConfig().CacheTreeHash {
```

File: shared/ssz/hash_cache_test.go [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+		if err := cache.AddRoot(bytesutil.ToBytes32(bytesutil.Bytes4(i)), []byte{1}); err != nil {
+			t.Fatal(err)
+		}
+	}
+	log.Printf(
+		"hash cache key size is %d, itemcount is %d",
+		maxCacheSize,
+		cache.hashCache.ItemCount(),
+	)
+	time.Sleep(time.Second)
```
Comment (prestonvanloon) [inline shared/ssz/hash_cache_test.go]:
```suggestion
	time.Sleep(1* time.Second)
```

I know it is the same behavior, but it reads better.

File: shared/ssz/hash_cache_test.go [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+
+	exists, fetchedInfo, err := cache.RootByEncodedHash(bytesutil.ToBytes32(hs))
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !exists {
+		t.Error("Expected blockInfo to exist")
+	}
+	if bytes.Compare(mr, fetchedInfo.MarkleRoot) != 0 {
```
Comment (prestonvanloon) [inline shared/ssz/hash_cache_test.go]:
```suggestion
	if !bytes.Equal(mr, fetchedInfo.MarkleRoot)  {
```

File: shared/ssz/hash_cache.go [hunk 24]
```diff
@@ -0,0 +1,248 @@
...
+	} else {
+		sszUtils, err := cachedSSZUtils(rval.Type())
+		if err != nil {
+			return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+		}
+		output, err := sszUtils.hasher(rval)
+		if err != nil {
+			return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+		}
+		// Right-pad with 0 to make 32 bytes long, if necessary
```
Comment (prestonvanloon) [inline shared/ssz/hash_cache.go]:
```suggestion
		// Right-pad with 0 to make 32 bytes long, if necessary.
```

File: shared/ssz/hash.go [hunk 25]
```diff
@@ -6,13 +6,17 @@ import (
 	"fmt"
 	"reflect"
 
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
+
```
Comment (prestonvanloon) [inline shared/ssz/hash.go]: remove this space please

File: shared/ssz/hash_cache.go [hunk 26]
```diff
@@ -0,0 +1,248 @@
...
+	"github.com/karlseguin/ccache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+var (
+	// ErrNotMerkleRoot will be returned when a cache object is not a merkle root
```
Comment (prestonvanloon) [inline shared/ssz/hash_cache.go]:
```suggestion
	// ErrNotMerkleRoot will be returned when a cache object is not a merkle root.
```

File: shared/ssz/hash_cache.go [hunk 27]
```diff
@@ -0,0 +1,248 @@
...
+
+// hashCacheS struct with one queue for looking up by hash.
+type hashCacheS struct {
+	hashCache *ccache.Cache
+}
+
+// markleRoot specifies the hash of data in a struct
+type root struct {
+	Hash       common.Hash
+	MarkleRoot []byte
```
Comment (prestonvanloon) [inline shared/ssz/hash_cache.go]: I think this should be MerkleRoot? The concept was invented by Ralph Merkle.

File: shared/featureconfig/flags.go (hunk 20 above)
Comment (shayzluf) [reply shared/featureconfig/flags.go]: thanks

File: shared/ssz/hash.go (hunk 21 above)
Comment (shayzluf) [reply shared/ssz/hash.go]: done

File: shared/ssz/hash_cache_test.go (hunk 23 above)
Comment (shayzluf) [reply shared/ssz/hash_cache_test.go]: thanks

File: shared/featureconfig/config.go [hunk 28]
```diff
@@ -29,6 +29,7 @@ type FeatureFlagConfig struct {
 	EnableComputeStateRoot       bool // EnableComputeStateRoot implementation on server side.
 	EnableCrosslinks             bool // EnableCrosslinks in epoch processing.
 	EnableHistoricalStatePruning bool // EnableHistoricalStatePruning when updatifinalized states.
+	CacheTreeHash                bool // CacheTreeHash determent whether tree hashes will be cached.
```
Comment (terenc3t) [inline shared/featureconfig/config.go]:
```suggestion
	CacheTreeHash                bool // CacheTreeHash determines whether tree hashes will be cached.
```

File: shared/featureconfig/config.go [hunk 29]
```diff
@@ -59,6 +60,10 @@ func ConfigureBeaconFeatures(ctx *cli.Context) {
 		log.Info("Enabled crosslink computations")
 		cfg.EnableCrosslinks = true
 	}
+	if ctx.GlobalBool(CacheTreeHashFlag.Name) {
+		log.Info("Cache tree hashes for ssz.")
```
Comment (terenc3t) [inline shared/featureconfig/config.go]:
```suggestion
		log.Info("Cache tree hashes for ssz")
```

File: shared/featureconfig/config.go [hunk 30]
```diff
@@ -75,6 +80,10 @@ func ConfigureValidatorFeatures(ctx *cli.Context) {
 		log.Info("Verifying signatures for attestations")
 		cfg.VerifyAttestationSigs = true
 	}
+	if ctx.GlobalBool(CacheTreeHashFlag.Name) {
+		log.Info("Cache tree hashes for ssz.")
```
Comment (terenc3t) [inline shared/featureconfig/config.go]:
```suggestion
		log.Info("Cache tree hashes for ssz")
```

File: shared/params/config.go [hunk 31]
```diff
@@ -23,6 +23,7 @@ type BeaconChainConfig struct {
 	WithdrawalPrivkeyFileName    string // WithdrawalPrivKeyFileName specifies the string name of a withdrawal private key file.
 	BLSPubkeyLength              int    // BLSPubkeyLength defines the expected length of BLS public keys in bytes.
 	DefaultBufferSize            int    // DefaultBufferSize for channels across the Prysm repository.
+	HashCacheSize                int64  // HashCacheSize defines the size of object hashes that are cached
```
Comment (terenc3t) [inline shared/params/config.go]:
```suggestion
	HashCacheSize                int64  // HashCacheSize defines the size of object hashes that are cached.
```

File: shared/ssz/hash_cache.go [hunk 32]
```diff
@@ -0,0 +1,248 @@
...
+var (
+	// ErrNotMerkleRoot will be returned when a cache object is not a merkle root.
+	ErrNotMerkleRoot = errors.New("object is not a merkle root")
+	// maxCacheSize is 2x of the follow distance for additional cache padding.
+	// Requests should be only accessing blocks within recent blocks within the
+	// Eth1FollowDistance.
+	maxCacheSize = params.BeaconConfig().HashCacheSize
+	// Metrics
+	hashCacheMiss = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "powchain_hash_cache_miss",
```
Comment (terenc3t) [inline shared/ssz/hash_cache.go]: why powchain?

File: shared/ssz/hash_cache.go [hunk 33]
```diff
@@ -0,0 +1,248 @@
...
+		Help: "The number of hash requests that aren't present in the cache.",
+	})
+	hashCacheHit = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "powchain_hash_cache_hit",
```
Comment (terenc3t) [inline shared/ssz/hash_cache.go]: why powchain?

File: shared/ssz/hash_cache.go [hunk 34]
```diff
@@ -0,0 +1,248 @@
...
+		Help: "The number of hash requests that are present in the cache.",
+	})
+	hashCacheSize = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "powchain_hash_cache_size",
```
Comment (terenc3t) [inline shared/ssz/hash_cache.go]: why powchain?

File: shared/ssz/hash_cache.go [hunk 35]
```diff
@@ -0,0 +1,248 @@
...
+	hashCache *ccache.Cache
+}
+
+// root specifies the hash of data in a struct
+type root struct {
+	Hash       common.Hash
+	MerkleRoot []byte
+}
+
+// hashKeyFn takes the  representation as the key for a hashInfo.
```
Comment (terenc3t) [inline shared/ssz/hash_cache.go]:
```suggestion
// hashKeyFn takes the representation as the key for a hashInfo.
```

File: shared/ssz/hash_cache.go [hunk 36]
```diff
@@ -0,0 +1,248 @@
...
+func newHashCache() *hashCacheS {
+	return &hashCacheS{
+		hashCache: ccache.New(ccache.Configure().MaxSize(maxCacheSize)),
+	}
+}
+
+// RootByEncodedHash fetches Root by the encoded hash of the object. Returns true with a
+// reference to the root if exists. Otherwise returns false, nil.
+func (b *hashCacheS) RootByEncodedHash(h common.Hash) (bool, *root, error) {
+
```
Comment (terenc3t) [inline shared/ssz/hash_cache.go]:
```suggestion
```

File: shared/ssz/hash_cache.go [hunk 37]
```diff
@@ -0,0 +1,248 @@
...
+func (b *hashCacheS) AddRoot(h common.Hash, rootB []byte) error {
+	mr := &root{
+		Hash:       h,
+		MerkleRoot: rootB,
+	}
+	b.hashCache.Set(mr.Hash.Hex(), mr, time.Hour)
+	return nil
+}
+
+// MakeSliceHasherCache add caching mechanism to slice hasher
```
Comment (terenc3t) [inline shared/ssz/hash_cache.go]:
```suggestion
// MakeSliceHasherCache add caching mechanism to slice hasher.
```

File: shared/ssz/hash_cache.go (hunk 34 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: you are right. no connection to powchain

File: shared/ssz/hash_cache.go [hunk 38]
```diff
@@ -0,0 +1,247 @@
...
+		Help: "The number of hashes in the block cache",
+	})
+)
+
+// hashCacheS struct with one queue for looking up by hash.
+type hashCacheS struct {
+	hashCache *ccache.Cache
+}
+
+// root specifies the hash of data in a struct
```
Comment (terenc3t) [inline shared/ssz/hash_cache.go]:
```suggestion
// root specifies the hash of data in a struct.
```

File: shared/ssz/hash_cache.go [hunk 39]
```diff
@@ -0,0 +1,247 @@
...
+)
+
+var (
+	// ErrNotMerkleRoot will be returned when a cache object is not a merkle root.
+	ErrNotMerkleRoot = errors.New("object is not a merkle root")
+	// maxCacheSize is 2x of the follow distance for additional cache padding.
+	// Requests should be only accessing blocks within recent blocks within the
+	// Eth1FollowDistance.
+	maxCacheSize = params.BeaconConfig().HashCacheSize
+	// Metrics
```
Comment (terenc3t) [inline shared/ssz/hash_cache.go]:
```suggestion
	// Metrics.
```

File: shared/ssz/hash_cache.go [hunk 40]
```diff
@@ -0,0 +1,247 @@
...
+		err = b.AddRoot(bytesutil.ToBytes32(hs), paddedOutput[:])
+		if err != nil {
+			return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+		}
+	}
+
+	return paddedOutput, nil
+}
+
+// MerkleHashCached adds a mrakle object to the cache. This method also trims the
```
Comment (nisdas) [inline shared/ssz/hash_cache.go]:
```suggestion
// MerkleHashCached adds a merkle object to the cache. This method also trims the
```

File: shared/ssz/hash_cache_test.go [hunk 41]
```diff
@@ -0,0 +1,182 @@
...
+	Int64Slice   []uint64
+}
+
+type tree struct {
+	First  []*JunkObject
+	Second []*JunkObject
+}
+
+// GenerateJunkObject generates junk object.
+func GenerateJunkObject(size uint64) []*JunkObject {
```
Comment (nisdas) [inline shared/ssz/hash_cache_test.go]: why does a test function this need to be exported ?

File: shared/ssz/hash_cache.go [hunk 42]
```diff
@@ -0,0 +1,247 @@
...
+	hasher := func(val reflect.Value) ([]byte, error) {
+		hs, err := hashedEncoding(val)
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode element of slice/array: %v", err)
+		}
+		exists, fetchedInfo, err := hashCache.RootByEncodedHash(bytesutil.ToBytes32(hs))
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode element of slice/array: %v", err)
+		}
+		var result [32]byte
```
Comment (nisdas) [inline shared/ssz/hash_cache.go]: you can write it this way
```go
             if exists {
			return bytesutil.ToBytes32(fetchedInfo.MerkleRoot)
		}
			concatElemHash := make([]byte, 0)
			for _, f := range fields {
				elemHash, err := f.sszUtils.hasher(val.Field(f.index))
				if err != nil {
					return nil, fmt.Errorf("failed to hash field of struct: %v", err)
				}
				concatElemHash = append(concatElemHash, elemHash...)
			}
		return hashutil.Hash(concatElemHash), nil
```

File: shared/ssz/hash_cache.go [hunk 43]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+// root specifies the hash of data in a struct.
+type root struct {
+	Hash       common.Hash
+	MerkleRoot []byte
+}
+
+// hashKeyFn takes the representation as the key for a hashInfo.
+func hashKeyFn(obj interface{}) (string, error) {
```
Comment (rauljordan) [inline shared/ssz/hash_cache.go]: Unused func

File: shared/ssz/hash_cache.go (hunk 40 above)
Comment (rauljordan) [reply shared/ssz/hash_cache.go]: +1

File: shared/ssz/hash_cache_test.go [hunk 44]
```diff
@@ -0,0 +1,182 @@
...
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
+)
+
+type JunkObject struct {
```
Comment (rauljordan) [inline shared/ssz/hash_cache_test.go]: No need to export this as well

File: shared/ssz/hash_cache_test.go [hunk 45]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+func TestMerkleHashWithCache(t *testing.T) {
+	cache := newHashCache()
+	for i := 0; i < 200; i++ {
+
+		runMerkleHashTests(t, func(val [][]byte) ([]byte, error) {
+			return merkleHash(val)
+		})
+
```
Comment (rauljordan) [inline shared/ssz/hash_cache_test.go]: Remove unnecessary blank lines

File: shared/ssz/hash_cache_test.go [hunk 46]
```diff
@@ -0,0 +1,182 @@
...
+		if err := cache.AddRoot(bytesutil.ToBytes32(bytesutil.Bytes4(i)), []byte{1}); err != nil {
+			t.Fatal(err)
+		}
+	}
+	log.Printf(
+		"hash cache key size is %d, itemcount is %d",
+		maxCacheSize,
+		cache.hashCache.ItemCount(),
+	)
+	time.Sleep(1 * time.Second)
```
Comment (rauljordan) [inline shared/ssz/hash_cache_test.go]: Why sleep here?

Comment (shayzluf) [reply shared/ssz/hash_cache_test.go]: there is some race condition on the go-routine causes the test to finish before the goroutine exited

File: shared/ssz/hash_cache.go (hunk 43 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: you are right. stayed there from using a different cache method.
thanks

File: shared/ssz/hash_cache.go (hunk 40 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: fixed

File: shared/ssz/hash_cache.go (hunk 42 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: thanks

File: shared/ssz/hash_cache_test.go (hunk 44 above)
Comment (shayzluf) [reply shared/ssz/hash_cache_test.go]: right

File: shared/ssz/hash_cache_test.go (hunk 45 above)
Comment (shayzluf) [reply shared/ssz/hash_cache_test.go]: done

**Comment by rauljordan** [comment]
@nisdas PTAL

**Comment by nisdas** [comment]
@shayzluf please remove the debug binary

File: beacon-chain/core/helpers/deposits_test.go [hunk 47]
```diff
@@ -6,9 +6,16 @@ import (
 
 	"github.com/gogo/protobuf/proto"
 	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
 	"github.com/prysmaticlabs/prysm/shared/params"
 )
 
+func init() {
+	featureconfig.InitFeatureConfig(&featureconfig.FeatureFlagConfig{
```
Comment (nisdas) [inline beacon-chain/core/helpers/deposits_test.go]: are all these inits required ? isnt it false by default

File: shared/ssz/hash_cache.go [hunk 48]
```diff
@@ -0,0 +1,234 @@
...
+// This method also trims the least recently added root info. if the cache size
+// has reached the max cache size limit.
+func (b *hashCacheS) TrieRootCached(val interface{}) ([32]byte, error) {
+	if val == nil {
+		return [32]byte{}, newHashError("untyped nil is not supported", nil)
+	}
+	rval := reflect.ValueOf(val)
+	hs, err := hashedEncoding(rval)
+	if err != nil {
+		return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
```
Comment (nisdas) [inline shared/ssz/hash_cache.go]: cant we just use `err.Error()` and get the string ? no need to use `fmt.Sprint`

File: shared/ssz/hash_cache_test.go (hunk 46 above)
Comment (nisdas) [reply shared/ssz/hash_cache_test.go]: we should not be using sleep in tests, it makes them brittle, anyway we can use a condition in a for loop , to wait for it. Or use channels to signify that the gorutine has finished

File: beacon-chain/core/helpers/deposits_test.go (hunk 47 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/deposits_test.go]: I get panic if i don't init them

File: shared/ssz/hash_cache.go (hunk 48 above)
Comment (shayzluf) [reply shared/ssz/hash_cache.go]: done

**Comment by nisdas** [comment]
@shayzluf please remove this file 
https://github.com/prysmaticlabs/prysm/blob/f0b11b08b9ced99a4704ba7ff165bccd9538436c/beacon-chain/debug

**Comment by shayzluf** [comment]
done
